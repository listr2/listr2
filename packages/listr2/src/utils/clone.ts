import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
