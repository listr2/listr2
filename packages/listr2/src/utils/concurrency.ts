// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
