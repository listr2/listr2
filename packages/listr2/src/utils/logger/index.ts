export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
