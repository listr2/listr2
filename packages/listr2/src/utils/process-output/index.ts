export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
