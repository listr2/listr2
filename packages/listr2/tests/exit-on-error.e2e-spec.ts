import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('exit on error', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  it('should throw out an error when exit on error is disabled', async() => {
    const task = new Listr(
      [
        {
          task: (): void => {
            throw new Error('failed')
          }
        },
        {
          task: (): void => {
            throw new Error('failed')
          }
        }
      ],
      { renderer: 'test', exitOnError: true }
    )

    let result: any
    let crash: Error

    try {
      result = await task.run()
    } catch(e: any) {
      crash = e
    }

    expect(crash).toBeTruthy()
    expect(result).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, 'LzVmO4zGsDmmp1zbNsmTMjuiKEGFFPSM')
  })

  it('should contunie execution while exit on error is disabled', async() => {
    const task = new Listr(
      [
        {
          task: (): void => {
            throw new Error('failed')
          }
        },
        {
          task: (): void => {
            throw new Error('failed')
          }
        }
      ],
      { renderer: 'test', exitOnError: false }
    )

    const result = await task.run()

    expect(result).toBeTruthy()
    expectProcessOutputToMatchSnapshot(output, 'EjYNt9Z2RtF4tqNsJ5PmTfYGhP6R70OH')
  })

  it('should contunie execution while exit on error is disabled in subtask', async() => {
    const task = new Listr(
      [
        {
          task: (): Listr =>
            new Listr(
              [
                {
                  task: (): void => {
                    throw new Error('failed')
                  }
                }
              ],
              { exitOnError: false }
            )
        }
      ],
      { renderer: 'test', exitOnError: true }
    )

    const result = await task.run()

    expect(result).toBeTruthy()
    expectProcessOutputToMatchSnapshot(output, 'MNdM4qgvgd0A43ALVQPcEa8EBq7wHja3')
  })

  it('should contunie execution while exit on error is disabled in parenttask', async() => {
    const task = new Listr(
      [
        {
          task: (): Listr =>
            new Listr(
              [
                {
                  task: (): void => {
                    throw new Error('failed')
                  }
                }
              ],
              { exitOnError: true }
            )
        }
      ],
      { renderer: 'test', exitOnError: false }
    )

    const result = await task.run()

    expect(result).toBeTruthy()
  })

  it('should throw out an error if subtask fails while exit on error is disabled', async() => {
    const task = new Listr(
      [
        {
          task: (): Listr =>
            new Listr(
              [
                {
                  task: (): void => {
                    throw new Error('failed')
                  }
                }
              ],
              { exitOnError: true }
            )
        }
      ],
      { renderer: 'test', exitOnError: true }
    )

    let result: any
    let crash: Error

    try {
      result = await task.run()
    } catch(e: any) {
      crash = e
    }

    expect(crash).toBeTruthy()
    expect(result).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, '4R3XUruBArL7ZBSS1X6VlufOYH9cSAo8')
  })

  it('should throw out an error if subtask in subtask fails while exit on error is disabled', async() => {
    const task = new Listr(
      [
        {
          task: (): Listr =>
            new Listr(
              [
                {
                  task: (): Listr =>
                    new Listr([
                      {
                        task: (): void => {
                          throw new Error('failed')
                        }
                      }
                    ])
                }
              ],
              { exitOnError: true }
            )
        }
      ],
      { renderer: 'test', exitOnError: true }
    )

    let result: any
    let crash: Error

    try {
      result = await task.run()
    } catch(e: any) {
      crash = e
    }

    expect(crash).toBeTruthy()
    expect(result).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, 'cAcT7aLZ2rDDFDH6EllJjYaG4mTe7Rrb')
  })
})
