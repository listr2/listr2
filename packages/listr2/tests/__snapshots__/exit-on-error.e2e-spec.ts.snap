// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`exit on error should contunie execution while exit on error is disabled in subtask: MNdM4qgvgd0A43ALVQPcEa8EBq7wHja3-exit 1`] = `[]`;

exports[`exit on error should contunie execution while exit on error is disabled in subtask: MNdM4qgvgd0A43ALVQPcEa8EBq7wHja3-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should contunie execution while exit on error is disabled in subtask: MNdM4qgvgd0A43ALVQPcEa8EBq7wHja3-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should contunie execution while exit on error is disabled: EjYNt9Z2RtF4tqNsJ5PmTfYGhP6R70OH-exit 1`] = `[]`;

exports[`exit on error should contunie execution while exit on error is disabled: EjYNt9Z2RtF4tqNsJ5PmTfYGhP6R70OH-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should contunie execution while exit on error is disabled: EjYNt9Z2RtF4tqNsJ5PmTfYGhP6R70OH-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error if subtask fails while exit on error is disabled: 4R3XUruBArL7ZBSS1X6VlufOYH9cSAo8-exit 1`] = `[]`;

exports[`exit on error should throw out an error if subtask fails while exit on error is disabled: 4R3XUruBArL7ZBSS1X6VlufOYH9cSAo8-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error if subtask fails while exit on error is disabled: 4R3XUruBArL7ZBSS1X6VlufOYH9cSAo8-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error if subtask in subtask fails while exit on error is disabled: cAcT7aLZ2rDDFDH6EllJjYaG4mTe7Rrb-exit 1`] = `[]`;

exports[`exit on error should throw out an error if subtask in subtask fails while exit on error is disabled: cAcT7aLZ2rDDFDH6EllJjYaG4mTe7Rrb-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error if subtask in subtask fails while exit on error is disabled: cAcT7aLZ2rDDFDH6EllJjYaG4mTe7Rrb-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error when exit on error is disabled: LzVmO4zGsDmmp1zbNsmTMjuiKEGFFPSM-exit 1`] = `[]`;

exports[`exit on error should throw out an error when exit on error is disabled: LzVmO4zGsDmmp1zbNsmTMjuiKEGFFPSM-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"failed"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;

exports[`exit on error should throw out an error when exit on error is disabled: LzVmO4zGsDmmp1zbNsmTMjuiKEGFFPSM-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"hasReset":false,"hasTitle":false,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":[null]}}
",
]
`;
